homeassistant/components/recorder/purge.py

```diff
@@ -7,32 +7,63 @@
 import homeassistant.util.dt as dt_util
 
 from .models import Events, RecorderRuns, States
-from .util import session_scope
+from .util import execute, session_scope
 
 _LOGGER = logging.getLogger(__name__)
 
 
-def purge_old_data(instance, purge_days, repack):
-    """Purge events and states older than purge_days ago."""
+def purge_old_data(instance, purge_days: int, repack: bool) -> bool:
+    """Purge events and states older than purge_days ago.
+
+    Cleans up an timeframe of an hour, based on the oldest record.
+    """
     purge_before = dt_util.utcnow() - timedelta(days=purge_days)
     _LOGGER.debug("Purging events before %s", purge_before)
 
     try:
         with session_scope(session=instance.get_session()) as session:
-            deleted_rows = (
+            query = session.query(States).order_by(States.last_updated.asc()).limit(1)
+            states = execute(query, to_native=True, validate_entity_ids=False)
+
+            states_purge_before = purge_before
+            if states:
+                states_purge_before = min(
+                    purge_before, states[0].last_updated + timedelta(hours=1)
+                )
+
+            deleted_rows_states = (
                 session.query(States)
-                .filter(States.last_updated < purge_before)
+                .filter(States.last_updated < states_purge_before)
                 .delete(synchronize_session=False)
             )
-            _LOGGER.debug("Deleted %s states", deleted_rows)
+            _LOGGER.debug("Deleted %s states", deleted_rows_states)
 
-            deleted_rows = (
+            query = session.query(Events).order_by(Events.time_fired.asc()).limit(1)
+            events = execute(query, to_native=True)
+
+            events_purge_before = purge_before
+            if events:
+                events_purge_before = min(
+                    purge_before, events[0].time_fired + timedelta(hours=1)
+                )
+
+            deleted_rows_events = (
                 session.query(Events)
-                .filter(Events.time_fired < purge_before)
+                .filter(Events.time_fired < events_purge_before)
                 .delete(synchronize_session=False)
             )
-            _LOGGER.debug("Deleted %s events", deleted_rows)
+            _LOGGER.debug("Deleted %s events", deleted_rows_events)
+
+        # If states or events purging isn't processing the purge_before yet,
+        # return false, as we are not done yet.
+        if (states_purge_before and states_purge_before != purge_before) or (
+            events_purge_before and events_purge_before != purge_before
+        ):
+            _LOGGER.debug("Purging hasn't fully completed yet.")
+            return False
 
+        # Recorder runs is small, no need to batch run it
+        with session_scope(session=instance.get_session()) as session:
```

Comment:
Can we indent the `if` above and avoid creating a second session scope?

Reply:
👍

Reply:
You are right, leftover from one of the many approaches I tried. Removed it.